source/_posts/angular-signals.md

```diff
@@ -0,0 +1,409 @@
+
+    ```ts
+    const counter = signal(0);
+    counter.set(1);
+    ```
+
+2. `update()` will update the value of the signal based on its current (old) value.
+
+    ```ts
+    const counter = signal(0);
+    counter.update(value => value + 1);
+    ```
+
+3. `mutate()` will mutate the current value internally and it is a sugar for `update()`. If we didn't want to do this: `users.update(users => [...users, newUser])`, then we can mutate if we want. It's great for objects.
+    ```ts
+    const users = signal<User[])([]);
+    
+    users.mutate(user => {
+        user.push({name: 'Lucas', age: 33});
+    });
+    ```
+
+Keep in mind that each of the above methods will also notify each dependency and its value where the signal was used.
+
+### Equality function
+
+It is possible to optionally add an equality comparator function to the signal. This is to check newly supplied values and whether they are the same, or different, compared to the current value.
+
+```ts
+const counter = signal(0, myEqualityFunction);
+```
+
+`myEqualityFunction` is type of: `ValueEqualityFn<T> = (a: T, b: T) => boolean`
+
+If the function detects that 2 values are the same, the signal value is not updated and the propagation of all changes in the dependencies of this signal is blocked. This gives you more control over updates in your signal.
+
+### What is computed()?
+
+It is a function that creates a memoizing computed signal, which **returns a reactive value from an expression**.
+
+```ts
+const counter = signal(0);
+const isEqual = computed(() => counter() === 5);
+```
+
+`isEqual` calculates its value by using `computed()` from signal value. It automatically updates the result every time the value of the `counter` signal changes.
+
+The important thing to remember is that each signal reading within `computed()` is *treated as a dependency*, and it will be tracked in the dependency graph.
+
+```ts
+const counter = signal(0);
+const isEqual = computed(() => counter() === 5); // counter() first dependency
+const isOdd = computed(() => counter() % 2); // counter() second dependency
+```
+
+And if we now change the signal value e.g. with `.set()`, the changes will be propagated in all `computed` functions, where the signal reading `counter()` is located. These locations are dependencies.
+
+{% img "ohh-yea.jpg" " " "lazy" %}
+
+It is worth mentioning that `computed()` values are cached. They are updated only when it’s needed.
+
+### Signal Effects
+
+`effect()` is a tree-shakable side effect function that **triggers when a signal value changes**. All signal dependencies are captured inside the function. Then, if there is a change in the signal, the side effect function is re-executed.
+
+```ts
+const counter = signal(0);
+
+effect(() => {
+    console.log('Counter value ', this.counter());
+
+    if (this.counter() === 5) {
+        console.log('Counter value is equal 5!');
+    }
+});
+
+
+counter.set(1);
+counter.set(2);
+counter.set(5);
+
+
+// Counter value 0
```

Comment:
Do we need 2 new lines?

Reply:
2x Double lines removed. But we need 1 line breaker to have better readability in code snippet